### PR TITLE
Split enrollments of the ValidatorSet into candidates and validators

### PR DESCRIPTION
**Pull request by linked0**
The old `ValidatorSet` had an issue that it manages two kinds of enrollments which consist of the enrollments waiting to be a validator and the enrollments having already been a validator. This could cause an unexpected error, so we are splitting the class into the `EnrollmentPoo`l for the candidates of a validator and the `ValidatorSet` for the confirmed validators.

**In this PR, I have just split the ValidatorSet class into two classes, EnrollmentPool and ValidatorSet for smaller PR.** 

There will be other PRs having the following modifications.
1. Change unnecessary public into private in the `ValidatorSet`
2. Change function names in the `ValidatorSet` into proper names like `getEnrollment` into `getValidator`
3. Check abnormal situations in `add` of the `ValidatorSet` and `updateEnrolledHeight` of the EnrollmentManager.

Relates #735

### CODE REVIEW

**Comment by linked0** [comment]
Ready for review.

File: source/agora/consensus/EnrollmentPool.d [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+            return false;
+        }
+
+        try
+        {
+            () @trusted {
+                this.db.execute("INSERT INTO enrollment_pool (key, val) VALUES (?, ?)",
+                    enroll.utxo_key.toString(), buffer);
+            }();
+
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentPool.d]:
```suggestion
```

File: source/agora/consensus/EnrollmentPool.d [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+        }
+
+        return true;
+    }
+
+    /***************************************************************************
+
+        Returns:
+            the number of enrollments being managed by the enrollment pool,
+            which includes both registered and un-registered enrollments.
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentPool.d]: In `EnrollmentManager.updateEnrolledHeight`, the enrollment is removed from the pool when the height is updated. I assume that when the height is updated it means the enrollment has been registered. So then the pool only contains unregistered enrollments, right?

File: source/agora/consensus/EnrollmentPool.d [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+
+    ***************************************************************************/
+
+    public this (string db_path)
+    {
+        this.db = Database(db_path);
+
+        // create the table for enrollment pool if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS enrollment_pool " ~
+            "(key TEXT PRIMARY KEY, val BLOC NOT NULL)");
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentPool.d]:
```suggestion
            "(key TEXT PRIMARY KEY, val BLOB NOT NULL)");
```

File: source/agora/consensus/EnrollmentManager.d [hunk 4]
```diff
@@ -22,6 +22,7 @@ import agora.consensus.data.Block;
 import agora.consensus.data.Enrollment;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d:22]: Typo in commit message: candicates => candidates

File: source/agora/consensus/EnrollmentManager.d [hunk 5]
```diff
@@ -212,7 +218,13 @@ public class EnrollmentManager
     public bool updateEnrolledHeight (const ref Hash enroll_hash,
         const size_t block_height) @safe
     {
-        if (!this.validator_set.updateEnrolledHeight(enroll_hash, block_height))
+        Enrollment enroll;
+        if (!this.enroll_pool.getEnrollment(enroll_hash, enroll))
+            return false;
+
+        this.enroll_pool.remove(enroll_hash);
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d]: There is a problem here:

- We take one enrollment from the pool
- We remove it from the pool
- We try to add it to the validator set

If the last step fails the enrollment no longer exists anywhere because it was removed from the pool. We should only call `this.enroll_pool.remove(enroll_hash);` if updating the enrolled height (adding it to the validator set) was successful.

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]

File: source/agora/consensus/EnrollmentManager.d (hunk 5 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d]: Thanks for your comment. In the case, the function returns `false` and the ledger will terminate a node with the `assert(0)` call. The validator set can be restored with the ledger when a node starts.

Comment (AndrejMitrovic) [reply source/agora/consensus/EnrollmentManager.d]: Then the assert should be moved inside of here. Otherwise this is what's called "leaky abstractions".

Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d]: It could be better for the ledger to have the role of determining whether a node terminates or not. But, I'll think about what you suggested.

Comment (AndrejMitrovic) [reply source/agora/consensus/EnrollmentManager.d]: If you want the Ledger to have the option to recover, then you need to make sure the EnrollmentManager is in a valid state.

That means to not remove the enrollment from the pool unless it was successfully added to the validator set.

Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d]: The information for validators is stored in a block and the ledger restores validators from the block. At the time when the `updateEnrolledHeight` is being called, the validators' information has already been stored in a block.

File: source/agora/consensus/ValidatorSet.d [hunk 6]
```diff
@@ -83,56 +80,40 @@ public class ValidatorSet
...
-                this.db.execute("INSERT INTO validator_set (key, val) VALUES (?, ?)",
-                    enroll.utxo_key.toString(), buffer);
-            }();
+            if (this.hasEnrollment(enroll.utxo_key))
+            {
+                log.info("Rejected already existing validator, Data was: {}",
+                    enroll);
+                return false;
+            }
+            else
```
Comment (Geod24) [inline source/agora/consensus/ValidatorSet.d]: No need for `else` if you `return`.

File: source/agora/test/EnrollmentManager.d [hunk 7]
```diff
@@ -26,71 +26,7 @@ import agora.consensus.EnrollmentManager;
 import agora.consensus.Genesis;
 import agora.test.Base;
 
-/// test enrollment process
```
Comment (Geod24) [inline source/agora/test/EnrollmentManager.d:33]: Why was this removed ?

File: source/agora/consensus/ValidatorSet.d [hunk 8]
```diff
@@ -788,7 +681,7 @@ unittest
 
     // make test blocks used for restoring validator set
     Block[] blocks;
-    ulong last_height = ValidatorSet.ValidatorCycle;
+    ulong last_height = Enrollment.ValidatorCycle;
```
Comment (Geod24) [inline source/agora/consensus/ValidatorSet.d:702]: There's a lot of diff coming from simply moving this constant. I think it should be its own commit.

File: source/agora/consensus/EnrollmentManager.d (hunk 5 above)
Comment (Geod24) [reply source/agora/consensus/EnrollmentManager.d]: > It could be better for the ledger to have the role of determining whether a node terminates or not.

Ideally, it would be better if the node would not terminate ;)

> At the time when the `updateEnrolledHeight` is being called, the validators' information has already been stored in a block.

Right, but we don't read the block here, we read the `enroll_pool`. And if the `enroll_pool` does not contain the enrollment, we bail out. No checking the block, no requesting other nodes.

Also, I didn't see where it was called from ?

File: source/agora/consensus/EnrollmentPool.d [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+                        :memory: was passed
+
+    ***************************************************************************/
+
+    public this (string db_path)
+    {
+        this.db = Database(db_path);
+
+        // create the table for enrollment pool if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS enrollment_pool " ~
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentPool.d:59]: Do we really need a table here ?

**Review by Geod24 — CHANGES_REQUESTED** [review]
A few comments, let's discuss IRL the rest for simplicity.

File: source/agora/test/EnrollmentManager.d (hunk 7 above)
Comment (linked0) [reply source/agora/test/EnrollmentManager.d:33]: The unit test code has been merged into the code below the comment `/// test for  enrollment process & revealing a pre-image periodically`. Changing the way the EnrollmentManager stores the data make the test code need to be changed and I found there was redundant test code, I removed the redundant code and merged the part of it into the other test code.

File: source/agora/consensus/ValidatorSet.d (hunk 8 above)
Comment (linked0) [reply source/agora/consensus/ValidatorSet.d:702]: I will do it. but in some commit of this PR, there will be two `ValidatorCycle` constants because the `EnrollmentPool` must reference it. In the end, the ValidatorCycle will only have existed in the `Enrollment` class.

File: source/agora/consensus/ValidatorSet.d (hunk 6 above)
Comment (linked0) [reply source/agora/consensus/ValidatorSet.d]: Please share your idea in more detail.

File: source/agora/consensus/EnrollmentPool.d (hunk 9 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentPool.d:59]: If we don't need this new table, I didn't need to start to resolve this issue. This PR and the issue related to this is all about creating a new table like enrollment_pool.

Comment (linked0) [reply source/agora/consensus/EnrollmentPool.d:59]: > No checking the block, no requesting other nodes.

You're right. We need some recovery code like requesting enrollments from other nodes. We need another issue for that and I'm sorry for the late catch.

File: source/agora/consensus/ValidatorSet.d (hunk 6 above)
Comment (AndrejMitrovic) [reply source/agora/consensus/ValidatorSet.d]:
```D
if (condition)
{
    do();
    return;
}
else
{
    something();
}
```

Is the same as:

```D
if (condition)
{
    do();
    return;
}

something();
```

File: source/agora/consensus/EnrollmentPool.d (hunk 9 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentPool.d:59]: > No checking the block, no requesting other nodes

I have created an issue for the problem. #755

**Comment by linked0** [comment]
Ready for review.
I resolved the comments and created a new issue for an issue that can't be handled in this PR.

File: source/agora/consensus/EnrollmentManager.d [hunk 10]
```diff
@@ -847,7 +859,6 @@ unittest
     assert(man.getEnrolledHeight(utxo_hash2) == 0);
     man.getUnregistered(enrolls);
     assert(enrolls.length == 1);
-    assert(man.count() == 2);  // has not changed
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d]: Why was the unittest removed instead of just changing the number that's tested?

File: source/agora/consensus/EnrollmentPool.d [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+        }
+
+        // check if already exists
+        if (this.hasEnrollment(enroll.utxo_key))
+        {
+            log.info("Rejected already existing enrollment, Data was: {}",
+                enroll);
+            return false;
+        }
+
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentPool.d:107]: The `buffer` should be declared here where it's initially used, not at the start of the function. We always try to keep declarations close to where they are initially used, it prevents accidental usage of uninitialized variables.

File: source/agora/consensus/EnrollmentPool.d [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+import d2sqlite3.database;
+import d2sqlite3.library;
+import d2sqlite3.results;
+import d2sqlite3.sqlite3;
+
+mixin AddLogger!();
+
+/// Ditto
+public class EnrollmentPool
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentPool.d:39]: There is an inconsistency in the API.

`hasEnrollment` doesn't throw, but `count` throws, `add` doesn't throw, but `remove` throws, and then `getEnrollment` throws and **also** returns a status flag.

These methods need to be consistent.

File: source/agora/consensus/EnrollmentManager.d [hunk 13]
```diff
@@ -130,44 +135,45 @@ public class EnrollmentManager
...
-        return this.validator_set.add(block_height, finder, enroll);
+        return this.enroll_pool.add(block_height, finder, enroll);
     }
 
     /***************************************************************************
 
         Returns:
-            the number of enrollments being managed by this EnrollmentManager.
+            the number of enrollments being managed by the enrollment pool.
             Note: this includes both registered and un-registered enrollments.
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d]: I think the Note: should be removed, because the pool only tracks un-registered enrollments.

**Review by AndrejMitrovic — DISMISSED** [review]

File: source/agora/consensus/EnrollmentPool.d [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+    ***************************************************************************/
+
+    public bool add (ulong block_height, scope UTXOFinder finder,
+        const ref Enrollment enroll) @safe nothrow
+    {
+        static ubyte[] buffer;
+
+        // check validity of the enrollment data
+        if (auto reason = isInvalidEnrollmentReason(enroll, block_height + 1,
+            finder))
```
Comment (linked0) [inline source/agora/consensus/EnrollmentPool.d]: The `block_height` is not used properly, the function of `isInvalidEnrollmentReason` should be changed.

File: source/agora/consensus/ValidatorSet.d [hunk 15]
```diff
@@ -83,56 +84,38 @@ public class ValidatorSet
...
-        }
-
         try
         {
+            if (this.hasEnrollment(enroll.utxo_key))
+            {
+                log.info("Rejected already existing validator, Data was: {}",
+                    enroll);
+                return false;
+            }
```
Comment (linked0) [inline source/agora/consensus/ValidatorSet.d]: This code doesn't need to be here in this `try/catch` statement.

File: source/agora/consensus/EnrollmentPool.d (hunk 12 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentPool.d:39]: > These methods need to be consistent.

Right, those functions should be consistent. I'll make it so.

File: source/agora/consensus/EnrollmentManager.d (hunk 10 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d]: > Why was the unittest removed instead of just changing the number that's tested?

This kind of deletion of code could be confusing. I'll restore it.

File: source/agora/consensus/ValidatorSet.d [hunk 16]
```diff
@@ -83,56 +84,38 @@ public class ValidatorSet
...
     {
         static ubyte[] buffer;
-
-        // check validity of the enrollment data
-        if (auto reason = isInvalidEnrollmentReason(enroll, block_height + 1,
-            finder))
-        {
-            log.info("Invalid enrollment data: {}, Data was: {}", reason, enroll);
-            return false;
-        }
```
Comment (linked0) [inline source/agora/consensus/ValidatorSet.d]: Revert it

**Comment by linked0** [comment]
It needs some more modification. I'll request a review after the modifications are done.

**Comment by linked0** [comment]
Ready for review about the code that Mathias and I pair programmed.

**Comment by Geod24** [comment]
Could you put the first commit in its own PR ?

**Comment by linked0** [comment]
> Could you put the first commit in its own PR ?

OK. I will do it.

File: source/agora/consensus/EnrollmentPool.d [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+        Params:
+            finder = the delegate to find UTXOs with
+            enroll = the enrollment data to add
+
+        Returns:
+            true if the enrollment data has been added to the enrollment pool
+
+    ***************************************************************************/
+
+    public bool add (scope UTXOFinder finder, const ref Enrollment enroll)
```
Comment (linked0) [inline source/agora/consensus/EnrollmentPool.d]: Invert params

**Review by linked0 — COMMENTED** [review]
Some suggestions.

File: source/agora/consensus/EnrollmentManager.d [hunk 18]
```diff
@@ -130,25 +135,26 @@ public class EnrollmentManager
...
+        Add a enrollment data to the enrollment pool
 
         Params:
-            finder = the delegate to find UTXOs with
             enroll = the enrollment data to add
+            finder = the delegate to find UTXOs with
 
         Returns:
-            true if the enrollment data has been added to the validator set
+            true if the enrollment data has been added to the the enrollment pool
```
Comment (TrustHenry) [inline source/agora/consensus/EnrollmentManager.d]:
```suggestion
            true if the enrollment data has been added to the enrollment pool
```

File: source/agora/consensus/ValidatorSet.d [hunk 19]
```diff
@@ -310,7 +262,7 @@ public class ValidatorSet
         {
             () @trusted {
                 auto results = this.db.execute("SELECT val FROM validator_set" ~
-                    " WHERE enrolled_height is not null");
+                    " ORDER BY key ASC");
```
Comment (TrustHenry) [inline source/agora/consensus/ValidatorSet.d:265]: > Get all the current validators

This change requires the document to be changed.
The meaning of sorted should be added.

**Review by TrustHenry — DISMISSED** [review]

File: source/agora/consensus/EnrollmentPool.d [hunk 20]
```diff
@@ -0,0 +1,404 @@
...
+        size_t cnt;
+        try
+        {
+            cnt = this.db.execute("SELECT count(*) FROM enrollment_pool").
+                oneValue!size_t;
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error on count");
+        }
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentPool.d]:
```suggestion
        try
        {
            return this.db.execute("SELECT count(*) FROM enrollment_pool").
                oneValue!size_t;
        }
        catch (Exception ex)
        {
            log.error("Database operation error on count");
            return 0;
        }
```

And the documentation needs to be updated to say 0 is returned in case of errors.

File: source/agora/consensus/EnrollmentPool.d [hunk 21]
```diff
@@ -0,0 +1,404 @@
...
+    public void remove (const ref Hash enroll_hash) @trusted nothrow
+    {
+        try
+        {
+            this.db.execute("DELETE FROM enrollment_pool WHERE key = ?",
+                enroll_hash.toString());
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error on remove");
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentPool.d:175]: When I said before that the API needs to be consistent I meant the code should do either one of the following:

A) Return a status code for every function which can fail. So the calling code can handle it.
B) Make all functions throw if they can fail. The calling code then must use a try/catch to handle it, or it can let the exception propagate upwards so other calling code handles it.
C) Catch any exceptions and halt the program with `assert(0)`.

Right now the code doesn't do any of these. Instead it catches the exception, logs it, and continues execution.

But what is the purpose of continuing execution like nothing bad happened? Will the database automatically fix itself on the next call? What is the expectation here?

Comment (linked0) [reply source/agora/consensus/EnrollmentPool.d:175]: The failure to remove an enrollment data from the enrollment pool is not so critical and so execution can continue. In this PR, if adding the enrollment to the validator set fails, the ledger makes the system terminate as follows.

```
foreach (enrollment; block.header.enrollments)
    if (!this.enroll_man.addValidator(enrollment,
            block.header.height, this.utxo_set.getUTXOFinder()))
        assert(0);
```

Mathias and I discussed that in the pair programming session.
The recovery from the database failure can be executed when a node start, which is not implemented yet and could be another issue.

**Comment by linked0** [comment]
Ready for review.
Additionally, I will make another PR about deleting the table named `node_enroll_data` used in the `EnrollmentManager` class soon like next Monday. The table manages only the `next_reveal_height` value which can be just set to the block height after the validator set is restored through the `restoreValidators` function.

**Comment by Geod24** [comment]
Rebased on v0.x.x

**Review by Geod24 — APPROVED** [review]
